pkg/yurtctl/cmd/revert/edgenode.go

```diff
@@ -0,0 +1,182 @@
+package revert
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	nodeutil "k8s.io/kubernetes/pkg/controller/util/node"
+
+	"github.com/alibaba/openyurt/pkg/projectinfo"
+	"github.com/alibaba/openyurt/pkg/yurtctl/constants"
+	enutil "github.com/alibaba/openyurt/pkg/yurtctl/util/edgenode"
+	kubeutil "github.com/alibaba/openyurt/pkg/yurtctl/util/kubernetes"
+)
+
+type RevertEdgeNodeOptions struct {
+	clientSet       *kubernetes.Clientset
+	kubeletSvcPath  string
+	openyurtDir     string
+	podMainfestPath string
+}
+
+func NewRevertEdgeNodeOptions() *RevertEdgeNodeOptions {
+	return &RevertEdgeNodeOptions{}
+}
+
+func NewRevertEdgeNodeCmd() *cobra.Command {
+	r := NewRevertEdgeNodeOptions()
+	cmd := &cobra.Command{
+		Use:   "edgenode",
+		Short: "reverts the yurt node to a kubernetes node",
+		Run: func(cmd *cobra.Command, _ []string) {
+			if err := r.Complete(cmd.Flags()); err != nil {
+				klog.Fatalf("fail to complete the revert edgenode option: %s", err)
+			}
+			if err := r.RunRevertEdgeNode(); err != nil {
+				klog.Fatalf("fail to revert the yurt node to a kubernetes node: %s", err)
+			}
+		},
+	}
+	return cmd
+}
+
+func (r *RevertEdgeNodeOptions) Complete(flags *pflag.FlagSet) (err error) {
+	r.clientSet, err = kubeutil.GenClientSet(flags)
+	if err != nil {
+		return err
+	}
+
+	kubeletSvcPath := os.Getenv("KUBELET_SVC")
+	if kubeletSvcPath == "" {
+		kubeletSvcPath = enutil.KubeletSvcPath
+	}
+	r.kubeletSvcPath = kubeletSvcPath
+
+	openyurtDir := os.Getenv("OPENYURT_DIR")
+	if openyurtDir == "" {
+		openyurtDir = enutil.OpenyurtDir
+	}
+	r.openyurtDir = openyurtDir
+
+	podMainfestPath := os.Getenv("STATIC_POD_PATH")
+	if podMainfestPath == "" {
+		podMainfestPath = enutil.StaticPodPath
+	}
+	r.podMainfestPath = podMainfestPath
+
+	return
+}
+
+func (r *RevertEdgeNodeOptions) RunRevertEdgeNode() (err error) {
+	// 1. check the server version
+	if err = kubeutil.ValidateServerVersion(r.clientSet); err != nil {
+		return
+	}
+	klog.V(4).Info("the server version is valid")
+
+	// 2. check the state of worker node and remove label from node
+	nodeName, err := enutil.GetNodeName()
+	if err != nil {
+		return err
+	}
+	node, err := r.clientSet.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	isEdgeNode, ok := node.Labels[projectinfo.GetEdgeWorkerLabelKey()]
+	if ok && isEdgeNode == "true" {
+		// 2.1. check the state of worker nodes
+		_, condition := nodeutil.GetNodeCondition(&node.Status, v1.NodeReady)
```

Comment:
maybe we need add node ready status check at the begin of edgenode convert and revert command.

Reply:
`yurtctl revert` has node ready status check at the beginning of the command.

`yurctl revert edgenode`  also has node status check. Before node status check, checking the server version and getting node resource are necessary. Does the check time need to be changed?

Reply:
We can check the node status before conversion/reversion, but the node can still fail when the conversion/reversion is ongoing. To eliminate this, we need to add a CRD/Controller to manage the OpenYurt cluster state, but that can be another new feature. I think we can keep the concurrent implementation just for now.

Reply:
+1 keep the current implementation